.github/workflows/build.yml

```
@@ -46,54 +66,148 @@ jobs:
+    needs: [prepare]
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      - name: Cache Docker layers
+        uses: actions/cache@v2
+        with:
+          path: ${{ env.BUILDX_CACHE_DIR }}
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
+      - name: Create dist directory
+        run: mkdir -p dist
+      - name: Build image
+        id: docker_build
+        uses: docker/build-push-action@v2
+        with:
+          build-args: |
+            VERSION=${{ needs.prepare.outputs.source_version }}
+          cache-from: type=local,src=${{ env.BUILDX_CACHE_DIR }}
+          cache-to: type=local,dest=${{ env.BUILDX_CACHE_DIR }}
+          context: .
+          file: ./Dockerfile
+          outputs: type=docker,dest=dist/image.tar
+          tags: ${{ env.IMAGE_NAME }}:latest  # not to be pushed
+          labels: "\
+            org.opencontainers.image.created=${{
+              needs.prepare.outputs.created }}
+
+            org.opencontainers.image.description=${{
+              fromJson(needs.prepare.outputs.repometa).description }}
+
+            org.opencontainers.image.licenses=${{
+              fromJson(needs.prepare.outputs.repometa).license.spdx_id }}
+
+            org.opencontainers.image.revision=${{ github.sha }}
+
+            org.opencontainers.image.source=${{
+              fromJson(needs.prepare.outputs.repometa).clone_url }}
+
+            org.opencontainers.image.title=${{
+              fromJson(needs.prepare.outputs.repometa).name }}
+
+            org.opencontainers.image.url=${{
+              fromJson(needs.prepare.outputs.repometa).html_url }}
+
+            org.opencontainers.image.version=${{
+              needs.prepare.outputs.source_version }}"
+      - name: Compress image
+        run: gzip dist/image.tar
       - name: Upload artifacts
         uses: actions/upload-artifact@v2
         with:
           name: dist
           path: dist
   test:
+    name: "Test image"
     runs-on: ubuntu-latest
     needs: [build]
     steps:
       - uses: actions/checkout@v2
       - uses: actions/setup-python@v2
         with:
           python-version: 3.9
-      - name: Cache testing environments
-        uses: actions/cache@v2
+      - name: Cache pip test requirements
+        uses: actions/cache@v1
         with:
           path: ${{ env.PIP_CACHE_DIR }}
-          key: "test-${{ runner.os }}-\
-            ${{ hashFiles('**/requirements-test.txt') }}-\
-            ${{ hashFiles('**/requirements.txt') }}"
+          key: "${{ runner.os }}-pip-test-\
+            ${{ hashFiles('**/requirements-test.txt') }}"
           restore-keys: |
-            test-${{ runner.os }}-
+            ${{ runner.os }}-pip-test-
+            ${{ runner.os }}-pip-
```

It's unclear to me why you used this old foo here:
```suggestion
            test-${{ runner.os }}-
```

---

Oh, mi scusi!